src/Compiler/TypedTree/CompilerGlobalState.fs

```diff
@@ -15,62 +18,31 @@ open FSharp.Compiler.Text
 /// It is made concurrency-safe since a global instance of the type is allocated in tast.fs, and it is good
 /// policy to make all globally-allocated objects concurrency safe in case future versions of the compiler
 /// are used to host multiple concurrent instances of compilation.
-type NiceNameGenerator() =
-
-    let lockObj = obj()
-    let basicNameCounts = Dictionary<string, int>(100)
-
+type NiceNameGenerator() =    
+    let basicNameCounts = ConcurrentDictionary<string,Ref<int>>(max Environment.ProcessorCount 1, 127)
+   
     member _.FreshCompilerGeneratedName (name, m: range) =
-      lock lockObj (fun () ->
         let basicName = GetBasicNameOfPossibleCompilerGeneratedName name
-        let n =
-            match basicNameCounts.TryGetValue basicName with
-            | true, count -> count
-            | _ -> 0
-        let nm = CompilerGeneratedNameSuffix basicName (string m.StartLine + (match n with 0 -> "" | n -> "-" + string n))
-        basicNameCounts[basicName] <- n + 1
-        nm)
-
-    member _.Reset () =
-      lock lockObj (fun () ->
-        basicNameCounts.Clear()
-      )
+        let countCell = basicNameCounts.GetOrAdd(basicName,fun k -> ref 0)
+        let count = Interlocked.Increment(countCell)
+        
+        CompilerGeneratedNameSuffix basicName (string m.StartLine + (match (count-1) with 0 -> "" | n -> "-" + string n))
 
 /// Generates compiler-generated names marked up with a source code location, but if given the same unique value then
 /// return precisely the same name. Each name generated also includes the StartLine number of the range passed in
 /// at the point of first generation.
 ///
 /// This type may be accessed concurrently, though in practice it is only used from the compilation thread.
 /// It is made concurrency-safe since a global instance of the type is allocated in tast.fs.
-type StableNiceNameGenerator() =
+type StableNiceNameGenerator() = 
 
-    let lockObj = obj()
-
-    let names = Dictionary<string * int64, string>(100)
-    let basicNameCounts = Dictionary<string, int>(100)
+    let niceNames = ConcurrentDictionary<string * int64, string>(max Environment.ProcessorCount 1, 127)
```

Comment:
127 is such a great number, it's always good to see it written down.  100 is a great number too.  Is there something we can use as a heuristic for names, do we know how big this dictionary gets during a compilation.   Do we know how big the benefit of pre-sizing the dictionary is?

Reply:
Not at this stage of the compiler process I am afraid. 
Neither the memory nor the speed are measurable - as a rule of thumb in muscle memory, I tend to choose prime numbers and prefer slight-oversize. But of course, here is not measurement available to provide heuristics.

=> could be a default growing one, yes.